client/state/happychat/middleware.js

```diff
@@ -0,0 +1,181 @@
+
+const actionOfType = type => () => ( { type } );
+
+const setChatConnecting = actionOfType( HAPPYCHAT_CONNECTING );
+const setChatConnected = actionOfType( HAPPYCHAT_CONNECTED );
+const setChatDisconnected = actionOfType( HAPPYCHAT_DISCONNECTED );
+
+const receiveChatEvent = event => ( { type: HAPPYCHAT_RECEIVE_EVENT, event } );
+
+const connectChat = ( { dispatch, getState } ) => {
+	const state = getState();
+	const user = getCurrentUser( state );
+	const connectionStatus = getHappychatConnectionStatus( state );
+
+	// if chat is already connected then do nothing
+	if ( connectionStatus !== 'disconnected' ) {
+		debug( 'chat already initialized' );
+		return;
+	}
+	debug( 'opening happychat for user', user, connectionStatus );
+
+	dispatch( setChatConnecting() );
+	// create new session id and get signed identity data for authenticating
+	connection
+	.on( 'connect', () => {
+		debug( 'set connected!' );
+		dispatch( setChatConnected() );
+	} )
+	.on( 'reconnect', () => dispatch( setChatConnecting() ) )
+	.on( 'disconnect', () => dispatch( setChatDisconnected() ) )
+	.on( 'message', event => dispatch( receiveChatEvent( event ) ) )
+	.on( 'status', status => {
+		debug( 'set chat status', status );
+		dispatch( setHappychatChatStatus( status ) );
+	} )
+	.on( 'accept', isAvailable => dispatch( setHappychatAvailable( isAvailable ) ) );
+
+	debug( 'opening chat session', connection );
+
+	startSession()
+	.then( ( { session_id } ) => sign( { user, session_id } ) )
+	.then( ( { jwt } ) => connection.open( user.ID, jwt ) )
+	.then(
+		() => debug( 'socket connection opened' ),
+		e => debug( 'failed to start happychat session', e, e.stack )
+	);
+};
+
+const sendTyping = throttle( message => {
+	connection.typing( message );
+}, 1000, { leading: true, trailing: false } );
+
+const onMessageChange = ( store, message ) => {
+	if ( ! isEmpty( message ) ) {
+		sendTyping( message );
+	} else {
+		connection.notTyping();
+	}
+};
+
+const sendMessage = ( { dispatch }, message ) => {
+	dispatch( updateChatMessage( '' ) );
+	connection.send( message );
+};
+
+const notifyWhenMinimized = ( { getState, dispatch } ) => {
+	if ( ! isHappychatOpen( getState() ) ) {
+		debug( 'display badge for chat' );
+		dispatch( setHappychatBadgeVisible() );
+	}
+};
+
+const receiveChatTranscript = ( messages, timestamp ) => ( {
+	type: HAPPYCHAT_TRANSCRIPT_RECEIVE, messages, timestamp
+} );
+
+const requestTranscript = ( { getState, dispatch } ) => {
+	const timestamp = getHappychatTranscriptTimestamp( getState() );
+	connection.transcript( timestamp ).then(
+		result => dispatch( receiveChatTranscript( result.messages, result.timestamp ) ),
+		e => debug( 'failed to get transcript', e )
+	);
+};
+
+export default ( store ) => {
+	return next => action => {
+		switch ( action.type ) {
+			case CURRENT_USER_ID_SET:
+				if ( isHappychatChatActive( store.getState() ) ) {
+					connectChat( store );
```

Comment:
The confusing part here is that we track two different chat statuses in the reducer.

1. [`connectionStatus`](https://github.com/Automattic/wp-calypso/blob/2f7ac4e050e59ada1e0047698c3a12a482b68b8b/client/state/happychat/reducer.js#L124): the state of the socket.io connection
2,[`chatStatus1](https://github.com/Automattic/wp-calypso/blob/2f7ac4e050e59ada1e0047698c3a12a482b68b8b/client/state/happychat/reducer.js#L153): the status of the user's "chat" (not network/socket related)

The [`isHappychatChatActive`](https://github.com/Automattic/wp-calypso/blob/2f7ac4e050e59ada1e0047698c3a12a482b68b8b/client/state/happychat/selectors.js#L35-L38) selector checks the value of `chatStatus` which is serialized in the reducer across browser refreshes which means it the user leaves and comes back to calypso when can tell it there were "in a chat" and open the socket again using `connectChat`.